flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -170,9 +180,13 @@ public void recycle(MemorySegment segment) {
 					throw new IllegalStateException("Buffer pool is destroyed.");
 				}
 
-				final MemorySegment segment = availableMemorySegments.poll(2, TimeUnit.SECONDS);
-				if (segment != null) {
-					segments.add(segment);
+				synchronized (availabilityHelper) {
```

Comment:
+1 for replacing the `ArrayBlockingQueue` with a custom lock here.

But maybe we can go one step further, and unify this locking object with `synchronized (factoryLock)`? Instead of taking two locks separately, we might be able to have just one? This would both simplify the code and improve the performance, however performance here probably doesn't matter that much, but looking at one method that synchronizes on two different objects is a bit confusing and triggers the question "why?"

For example, I think this particular method could be just wrapped as a whole in the `synchronized (availabilityHelper) {` block.

While doing so it might be possible to simplify the concurrency model even further, and replace `private volatile boolean isDestroyed;` with just `private boolean isDestroyed;`.

Reply:
Yes, I also had the similar feeling before. But not quite sure whether it is feasible to do that (I mean considering carefully whether has deadlock concern) and how many efforts to pay for that. But it is actually worth doing so if making sense.

Reply:
I'll check if we can sync on only one object. If there is no deadlock problem, it's more elegant to sync on only one object.

Reply:
There can be deadlock if we sync on only one object. The buffer request thread will first sync on the  local pool and then on the global pool. The buffer redistribution thread will first sync on the global pool and then on the local pools.

Reply:
If it is not very easy to integrate the locks atm, we might focus on this issue in a separate ticket future, although i agree the codes seem a bit ugly to have to separate locks.

Although i have not thought all the cases through, I guess it is not very easy to consider completely to avoid potential risks in short time.